core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -659,9 +659,10 @@ private[spark] class Executor(
         val accumUpdates = task.collectAccumulatorUpdates()
         val metricPeaks = metricsPoller.getTaskMetricPeaks(taskId)
         // TODO: do not serialize value twice
-        val directResult = new DirectTaskResult(valueBytes, accumUpdates, metricPeaks)
-        val serializedDirectResult = ser.serialize(directResult)
-        val resultSize = serializedDirectResult.limit()
+        val directResult = new DirectTaskResult(valueByteBuffer, accumUpdates, metricPeaks)
+        val serializedDirectResult = SerializerHelper.serializeToChunkedBuffer(ser, directResult,
+          valueByteBuffer.size)
```

Comment:
This topic is also related to the `estimatedSize >> 3` in the chunk size estimation. Since the estimated size is hard to be accurate anyway, so if we leave the chunk size to be exactly the `estimatedSize`, we're potentially wasting a large chunk (for example, the estimatedSize is 8,000, while the exact serialized size is 8,001, we're mostly wasting the last chunk of size 8000 and only contain 1 useful byte).

And dividing the chunks to size of `estimatedSize / k`,  we're having a reasonably lower upper bound of the overhead(we're only wasting the last chunk of size 1000)

Given this estimation heuristic, we don't have to be super accurate(actually hardly possible to be exact) so using `valueByteBuffer.size` here for simplicity

Reply:
As I mentioned above, get reasonable upper bound - we can always improve in future.
The main reason why I suggested improving for accumulators particularly is because typically rest of DirectTaskResult in much smaller than result value - so a reasonable overestimation there might help minimize need for two large buffers when one might do.
But this is mostly a nit - we can add a comment to the code, if we are unsure of the heurestics, to improve it for future.

The main reason for suggesting this is to reduce wastage at driver.

Reply:
Make senses, updated accordingly, thanks!